File: days/day_02/main.ml
open Core

type entry = {range: int*int; letter: char; password: string}
[@@deriving sexp_of]

let file = "input.txt"

let part_one entries =
	List.fold entries ~init:0 ~f:(fun total entry ->
		let count = 
			String.count entry.password ~f:(Char.(=) entry.letter)
		in
		let (min, max) = entry.range in
		if (count >= min) && (count <= max)
		then total+1
		else total
	)
;;


let part_two entries =
	List.fold entries ~init:0 ~f:(fun total entry ->
		let (a, b) = entry.range in
		let pw_list = String.to_list entry.password in
		if Bool.(<>)
			(Char.(=) (List.nth_exn pw_list (a-1)) entry.letter)
			(Char.(=) (List.nth_exn pw_list (b-1)) entry.letter)
		then total+1
		else total
	)
;;


let () = 
	let start_time = Unix.gettimeofday () in
	let input = In_channel.read_lines file in
	let entries = 
		input
		|> List.map ~f:(String.split ~on:' ')
		|> List.map ~f:(function
			| r :: l :: password :: [] ->
				let range_list = r
					|> String.split ~on:'-'
					|> List.map ~f:int_of_string
				in
				let range = (List.nth_exn range_list 0, List.nth_exn range_list 1) in
				let letter = Char.of_string (String.strip l ~drop:(Char.(=) ':')) in
				{range; letter; password}
			| _ -> failwith "nope"
		)
	in
	let part_one_ans = part_one entries in
	let part_two_ans = part_two entries in
	printf "Day  2 (%.3fs)\n" ((Unix.gettimeofday ()) -. start_time);
	printf "  Part 1: %d\n" part_one_ans;
	printf "  Part 2: %d\n" part_two_ans
;;

File: days/day_02/main.py
import time
import os


def part_one(pw_data):
	total_valid = 0

	for entry in pw_data:
		count = entry['password'].count(entry['letter'])
		[m, M] = entry['range']
		total_valid += count in range(m, M+1)

	return total_valid
		

def part_two(pw_data):
	total_valid = 0

	for entry in pw_data:
		[a,b] = entry['range']
		l = entry['letter']
		pw = entry['password']
		total_valid += (pw[a-1] == l)^(pw[b-1] == l)

	return total_valid
		

def main():
	start_time = time.time()

	with open(os.path.dirname(__file__) + '/input.txt', 'r') as data:
		raw_pw_data = [line.split(' ') for line in data.readlines()]
		pw_data = []
		
		for raw_entry in raw_pw_data:
			entry = {}
			entry['range'] = [int(x) for x in raw_entry[0].split('-')]
			entry['letter'] = raw_entry[1].strip(':')
			entry['password'] = raw_entry[2].strip('\n')
			pw_data.append(entry)

		part_one_ans = part_one(pw_data)
		part_two_ans = part_two(pw_data)

		print('Day  2 ({:,.3f}s)'.format(time.time()-start_time))
		print('  Part 1: {}'.format(part_one_ans))
		print('  Part 2: {}'.format(part_two_ans))
		

if __name__ == "__main__":
		main()